### ISSUE TITLE
Maximum call stack exceeded

### ISSUE THREAD

**Issue by zsoi** [data-forge-notebook/wiki#6]
I got some async code loaded up in DFN+ and after a decent amount of iterations the execution engine suddenly stops working with this error in the log:

```
[error] (editor-window/faf7d) evaluation-engine [stderr]: RangeError: Maximum call stack size exceeded
    at AsyncTracker.getCellIndexForOp ([...]\AppData\Local\Programs\data-forge-notebook\ts-build\evaluation-engine\index.js:216693:3245)
    at AsyncTracker.getCellIndexForOp ([...]\AppData\Local\Programs\data-forge-notebook\ts-build\evaluation-engine\index.js:216693:3465)
    at AsyncTracker.getCellIndexForOp ([...]\AppData\Local\Programs\data-forge-notebook\ts-build\evaluation-engine\index.js:216693:3465)
    at AsyncTracker.getCellIndexForOp ([...]\AppData\Local\Programs\data-forge-notebook\ts-build\evaluation-engine\index.js:216693:3465)
(continues...)
```

I'm not really sure if its the fault of my code because I don't see any actual recursion patterns, it's all just asynchronously iterating over asynchronously fetched data.

Happens in 1.10.0 and the previous version as well.

**Comment by ashleydavis** [comment]
Hi @zsoi, thanks for reporting the issue.

Are you able to give me an example notebook that I can use to reproduce the problem?

Just attach it to this issue, please.

Thanks!

**Comment by zsoi** [comment]
Sure, had to strip it way down but this reproduces the issue:

[execution-engine-stack-depletion-repro.zip](https://github.com/data-forge-notebook/wiki/files/5606697/execution-engine-stack-depletion-repro.zip)

**Comment by ashleydavis** [comment]
Thanks so much, I'll take a look during the week.

**Comment by ashleydavis** [comment]
Ok this is quite a complex example!

I suspect that Data-Forge Notebook doesn't support generators properly.

Is there anyway you can simplify the example? That's going to help me nail this issue down.

**Comment by zsoi** [comment]
I've been trying to further simplify it but it's becoming difficult as apparently the issue doesn't happen anymore when it's simplified too much. All I can currently say is that it's dependent on `totalCount`. The higher it gets, the higher the chance that the stack overflow happens.

**Comment by ashleydavis** [comment]
I think the problem must be caused by the scale of the code you are running. I'm going to try scaling up my simple example to see how I can push it before it breaks DFN!

I'm pleased to say that DFN supports generators and async generators, here's my test notebook:

![image](https://user-images.githubusercontent.com/625126/100684914-088e8b00-33c7-11eb-9c3b-6ab49fc0c001.png)

**Comment by ashleydavis** [comment]
Ok I'm using my test to push DFN to the limit!

I've set the timeout to 1 millisecond and the sequence to 10000.

By my estimation, the notebook should complete in 10 seconds, but it appears to be hanging.

I'll let it run for a bit and see if it completes.

**Comment by ashleydavis** [comment]
I think I've fixed the issue. I was able to reproduce the "max callstack exceeded" error by ramping up my small test to many iterations. 

There was a recursive function being called that recurses more for each new async operation in a chain of operations. I did a small optimization to this (by caching intermediate async operation ids) that allows the function to early out almost always after one call.

This solves the callstack problem and should make for better performance in the evaluation engine as well!

Thanks so much for reporting this issue.

The fix will be available in release 1.10.2.
